### ISSUE TITLE
Build failure in ROS 2 Rolling

### ISSUE THREAD

**Issue by nuclearsandwich** [ros-tooling/rosbag2_storage_mcap#76]
## Description

rosbag2_storage_mcap is failing to build on ROS 2 Rolling.
The first build failures began on Nov 3 and may coincide with the 0.5.0 release of rosbag2_storage_mcap (ros/rosdistro#35174)

## Failing build links

* https://build.ros2.org/view/Rbin_ujv8_uJv8/job/Rbin_ujv8_uJv8__rosbag2_storage_mcap__ubuntu_jammy_arm64__binary/53
* https://build.ros2.org/view/Rbin_uJ64/job/Rbin_uJ64__rosbag2_storage_mcap__ubuntu_jammy_amd64__binary/53/

## System (please complete the following information)
 - OS: Ubuntu Jammy AMD64 and ARM64
 - ROS 2 Distro: Rolling

Example build output:

```
00:01:46.341 [  8%] Building CXX object CMakeFiles/rosbag2_storage_mcap.dir/src/mcap_storage.cpp.o
00:01:46.351 /usr/lib/ccache/c++ -DRCUTILS_ENABLE_FAULT_INJECTION -DROSBAG2_STORAGE_MCAP_BUILDING_LIBRARY -DROSBAG2_STORAGE_MCAP_HAS_SET_READ_ORDER -DROSBAG2_STORAGE_MCAP_HAS_STORAGE_FILTER_TOPIC_REGEX -DROSBAG2_STORAGE_MCAP_HAS_STORAGE_OPTIONS -DROSBAG2_STORAGE_MCAP_HAS_YAML_HPP -DROSBAG2_STORAGE_MCAP_OVERRIDE_SEEK_METHOD -DROSBAG2_STORAGE_MCAP_WRITER_CREATES_DIRECTORY -D_TEST_RESOURCES_DIR_PATH=\"/tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/test/rosbag2_storage_mcap\" -Drosbag2_storage_mcap_EXPORTS -I/tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/include -isystem /opt/ros/rolling/include/mcap_vendor -isystem /opt/ros/rolling/include/pluginlib -isystem /opt/ros/rolling/include/rcutils -isystem /opt/ros/rolling/include/rosbag2_storage -isystem /opt/ros/rolling/include -isystem /opt/ros/rolling/include/ament_index_cpp -isystem /opt/ros/rolling/include/class_loader -isystem /opt/ros/rolling/include/rcpputils -g -O2 -ffile-prefix-map=/tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0=. -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Wall -Wextra -Wpedantic -MD -MT CMakeFiles/rosbag2_storage_mcap.dir/src/mcap_storage.cpp.o -MF CMakeFiles/rosbag2_storage_mcap.dir/src/mcap_storage.cpp.o.d -o CMakeFiles/rosbag2_storage_mcap.dir/src/mcap_storage.cpp.o -c /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp
00:01:46.353 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:181:46: error: ‘ReadOrder’ in namespace ‘rosbag2_storage’ does not name a type
00:01:49.736   181 |   void set_read_order(const rosbag2_storage::ReadOrder&) override;
00:01:49.736       |                                              ^~~~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:181:8: error: ‘void rosbag2_storage_plugins::MCAPStorage::set_read_order(const int&)’ marked ‘override’, but does not override
00:01:49.736   181 |   void set_read_order(const rosbag2_storage::ReadOrder&) override;
00:01:49.736       |        ^~~~~~~~~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:499:57: error: ‘ReadOrder’ in namespace ‘rosbag2_storage’ does not name a type
00:01:49.736   499 | void MCAPStorage::set_read_order(const rosbag2_storage::ReadOrder& read_order) {
00:01:49.736       |                                                         ^~~~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp: In member function ‘void rosbag2_storage_plugins::MCAPStorage::set_read_order(const int&)’:
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:501:22: error: request for member ‘sort_by’ in ‘read_order’, which is of non-class type ‘const int’
00:01:49.736   501 |   switch (read_order.sort_by) {
00:01:49.736       |                      ^~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:502:27: error: ‘rosbag2_storage::ReadOrder’ has not been declared
00:01:49.736   502 |     case rosbag2_storage::ReadOrder::ReceivedTimestamp:
00:01:49.736       |                           ^~~~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:503:22: error: request for member ‘reverse’ in ‘read_order’, which is of non-class type ‘const int’
00:01:49.736   503 |       if (read_order.reverse) {
00:01:49.736       |                      ^~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:509:27: error: ‘rosbag2_storage::ReadOrder’ has not been declared
00:01:49.736   509 |     case rosbag2_storage::ReadOrder::File:
00:01:49.736       |                           ^~~~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:510:23: error: request for member ‘reverse’ in ‘read_order’, which is of non-class type ‘const int’
00:01:49.736   510 |       if (!read_order.reverse) {
00:01:49.736       |                       ^~~~~~~
00:01:49.736 /tmp/binarydeb/ros-rolling-rosbag2-storage-mcap-0.5.0/src/mcap_storage.cpp:516:27: error: ‘rosbag2_storage::ReadOrder’ has not been declared
00:01:49.736   516 |     case rosbag2_storage::ReadOrder::PublishedTimestamp:
00:01:49.736       |                           ^~~~~~~~~
00:01:49.736 make[4]: *** [CMakeFiles/rosbag2_storage_mcap.dir/build.make:79: CMakeFiles/rosbag2_storage_mcap.dir/src/mcap_storage.cpp.o] Error 1
```

**Comment by asymingt** [comment]
I can confirm the same behavior on my side with a custom rolling. I think the underlying reason is that there is a 0.17.x and 0.18.x version, the first of which doesn't support read ordering. I think the fix is a simple one that involves updating the compile flags to accommodate both versions correctly.

I have posted a fix #78

**Comment by asymingt** [comment]
PR merged. Can we close this now?

**Comment by MichaelOrlov** [comment]
- Closing this as fixed in #78
